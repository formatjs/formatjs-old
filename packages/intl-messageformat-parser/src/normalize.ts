import {
  MessageFormatElement,
  isPluralElement,
  isLiteralElement,
  LiteralElement
} from './types';
import { parse } from './parser';

const PLURAL_HASHTAG_REGEX = /(^|[^\\])#/g;

/**
 * Whether to convert `#` in plural rule options
 * to `{var, number}`
 * @param el AST Element
 * @param pluralStack current plural stack
 */
export function normalizeHashtagInPlural(els: MessageFormatElement[]) {
  els.forEach(el => {
    // If we're encountering a plural el
    if (!isPluralElement(el)) {
      return;
    }
    // Go down the options and search for # in any literal element
    Object.keys(el.options).forEach(id => {
      const opt = el.options[id];
      // If we got a match, we have to split this
      // and inject a NumberElement in the middle
      let matchingLiteralElIndex = -1;
      let literalEl: LiteralElement | undefined = undefined;
      for (let i = 0; i < opt.value.length; i++) {
        const el = opt.value[i] as LiteralElement;
        if (isLiteralElement(el) && PLURAL_HASHTAG_REGEX.test(el.value)) {
          matchingLiteralElIndex = i;
          literalEl = el;
          break;
        }
      }

      if (literalEl) {
        const newValue = literalEl.value.replace(
          PLURAL_HASHTAG_REGEX,
          `$1{${el.value}, number}`
        );
        const newEls = parse(newValue);
        opt.value.splice(matchingLiteralElIndex, 1, ...newEls);
      }
      normalizeHashtagInPlural(opt.value);
    });
  });
}
