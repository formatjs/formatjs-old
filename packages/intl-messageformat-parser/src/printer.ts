/*
 * Copyright 2015, Yahoo Inc.
 * Copyrights licensed under the New BSD License.
 * See the accompanying LICENSE file for terms.
 */

import {
  MessageFormatElement,
  isLiteralElement,
  LiteralElement,
  ArgumentElement,
  DateElement,
  TimeElement,
  NumberElement,
  SelectElement,
  PluralElement,
  isSelectElement,
  isArgumentElement,
  isDateElement,
  isTimeElement,
  isNumberElement,
  isPluralElement,
  TYPE,
  Skeleton,
  SKELETON_TYPE,
  NumberSkeletonToken,
  DateTimeSkeleton,
  isPoundElement,
} from './types';

export function printAST(ast: MessageFormatElement[]): string {
  return doPrintAST(ast, false);
}

export function doPrintAST(
  ast: MessageFormatElement[],
  isInPlural: boolean
): string {
  const printedNodes = ast.map(el => {
    if (isLiteralElement(el)) {
      return printLiteralElement(el, isInPlural);
    }

    if (isArgumentElement(el)) {
      return printArgumentElement(el);
    }
    if (isDateElement(el) || isTimeElement(el) || isNumberElement(el)) {
      return printSimpleFormatElement(el);
    }

    if (isPluralElement(el)) {
      return printPluralElement(el);
    }

    if (isSelectElement(el)) {
      return printSelectElement(el);
    }

    if (isPoundElement(el)) {
      return '#';
    }
  });

  return printedNodes.join('');
}

function printEscapedMessage(message: string): string {
  return message.replace(/([{}](?:.*[{}])?)/su, `'$1'`);
}

function printLiteralElement({value}: LiteralElement, isInPlural: boolean) {
  const escaped = printEscapedMessage(value);
  return isInPlural ? escaped.replace('#', "'#'") : escaped;
}

function printArgumentElement({value}: ArgumentElement) {
  return `{${value}}`;
}

function printSimpleFormatElement(
  el: DateElement | TimeElement | NumberElement
) {
  return `{${el.value}, ${TYPE[el.type]}${
    el.style ? `, ${printArgumentStyle(el.style)}` : ''
  }}`;
}

function printNumberSkeletonToken(token: NumberSkeletonToken): string {
  const {stem, options} = token;
  return options.length === 0
    ? stem
    : `${stem}${options.map(o => `/${o}`).join('')}`;
}

function printArgumentStyle(style: string | Skeleton) {
  if (typeof style === 'string') {
    return printEscapedMessage(style);
  } else if (style.type === SKELETON_TYPE.dateTime) {
    return `::${printDateTimeSkeleton(style)}`;
  } else {
    return `::${style.tokens.map(printNumberSkeletonToken).join(' ')}`;
  }
}

export function printDateTimeSkeleton(style: DateTimeSkeleton): string {
  return style.pattern;
}

function printSelectElement(el: SelectElement) {
  const msg = [
    el.value,
    'select',
    Object.keys(el.options)
      .map(id => `${id}{${doPrintAST(el.options[id].value, false)}}`)
      .join(' '),
  ].join(',');
  return `{${msg}}`;
}

function printPluralElement(el: PluralElement) {
  const type = el.pluralType === 'cardinal' ? 'plural' : 'selectordinal';
  const msg = [
    el.value,
    type,
    [
      el.offset ? `offset:${el.offset}` : '',
      ...Object.keys(el.options).map(
        id => `${id}{${doPrintAST(el.options[id].value, true)}}`
      ),
    ]
      .filter(Boolean)
      .join(' '),
  ].join(',');
  return `{${msg}}`;
}
