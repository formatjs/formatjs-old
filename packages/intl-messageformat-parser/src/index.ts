import { parse as pegParse, IParseOptions } from './parser';
import { Options, MessageFormatElement } from './types';
import { normalizeHashtagInPlural } from './normalize';
export * from './types';
export * from './parser';

export type ParseOptions = Options & IParseOptions;

export function parse(
  input: string,
  opts?: ParseOptions
): MessageFormatElement[] {
  const els = pegParse(input, opts);
  if (!opts || opts.normalizeHashtagInPlural !== false) {
    normalizeHashtagInPlural(els);
  }
  return els;
}
