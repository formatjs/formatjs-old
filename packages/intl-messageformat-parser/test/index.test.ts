import {pegParse, ParseOptions} from '../src';
import {printAST} from '../src/printer';

function allTests(opts?: ParseOptions) {
  [
    'Hello, World!',
    'Hello, {name}!',
    'My name is {FIRST} {LAST}, age {age, number}, time {time, time}, date {date, date}.',
    '{num, number, percent}',
    '{count, time}',
    '{numPhotos, plural, =0{no photos} =1{one photo} other{# photos}}',
    'Foo {var1, plural, =0{# var1} other{{var2, plural, =0{# var2} other{# var2-other}} # other}}',
    '{floor, selectordinal, =0{ground} one{#st} two{#nd} few{#rd} other{#th}} floor',
    '{gender, select, female {woman} male {man} other {person}}',
    '   some random test   ',
    '{  num , number,percent  }',
    '{c, plural, =1 { {text} project} other { {text} projects}}',
    '{c, plural, offset:-2 =-1 { {text} project} other { {text} projects}}', // negative
    '{c, plural, =99 { {text} project} other { {text} projects}}',
    `'{'`,
    `'}'`,
    // Escaping "#" needs to be special-cased so it remains escaped so
    // the runtime doesn't replace it when in a `pluralFormat` option.
    `'#'`,
    /**
     * @see http://userguide.icu-project.org/formatpegParse/messages#TOC-Quoting-Escaping
     * @see https://github.com/formatjs/formatjs/issues/97
     */
    "This '{isn''t}' obvious",
    "'{name}'",
    'this is {count,plural,offset:1 one{{count, number} dog} other{{count, number} dogs}}',
    `{type, select,
      drop {
        {units, plural,
          one {# drop}
          other {# drops}
        }
      }
      teaspoon {
        {units, plural,
          one {# teaspoon}
          other {# teaspoons}
        }
      }
      tablespoon {
        {units, plural,
          one {# tablespoon}
          other {# tablespoons}
        }
      }
    }`,
    '{howMany, select,\
      one {{actor1}}\
      other {\
          {nExtraActors, plural,\
          one {{actor1} and {actor2}}\
          other {{actor1} and {nExtraActors} others}\
          }}\
      } shared this file.',
  ].forEach(mess => {
    const ast = pegParse(mess, opts);
    it(`can pegParse '${mess}'`, function() {
      expect(ast).toMatchSnapshot();
    });
    it(`can print AST from '${mess}'`, function() {
      expect(printAST(ast)).toMatchSnapshot();
    });
  });
}

describe('pegParse()', function() {
  allTests();
});

describe('pegParse({ captureLocation: true })', function() {
  allTests();
});
