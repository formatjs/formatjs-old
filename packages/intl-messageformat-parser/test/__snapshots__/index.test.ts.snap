// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse() can parse '   some random test   ' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "   some random test   ",
  },
]
`;

exports[`parse() can parse ''#'' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "'#'",
  },
]
`;

exports[`parse() can parse ''{'' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "{",
  },
]
`;

exports[`parse() can parse ''{name}'' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "{name}",
  },
]
`;

exports[`parse() can parse ''}'' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "}",
  },
]
`;

exports[`parse() can parse '{  num , number,percent  }' 1`] = `
Array [
  Object {
    "style": "percent",
    "type": 2,
    "value": "num",
  },
]
`;

exports[`parse() can parse '{c, plural, =1 { {text} project} other { {text} projects}}' 1`] = `
Array [
  Object {
    "offset": 0,
    "options": Object {
      "=1": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": " ",
          },
          Object {
            "type": 1,
            "value": "text",
          },
          Object {
            "type": 0,
            "value": " project",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": " ",
          },
          Object {
            "type": 1,
            "value": "text",
          },
          Object {
            "type": 0,
            "value": " projects",
          },
        ],
      },
    },
    "pluralType": "cardinal",
    "type": 6,
    "value": "c",
  },
]
`;

exports[`parse() can parse '{c, plural, =99 { {text} project} other { {text} projects}}' 1`] = `
Array [
  Object {
    "offset": 0,
    "options": Object {
      "=99": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": " ",
          },
          Object {
            "type": 1,
            "value": "text",
          },
          Object {
            "type": 0,
            "value": " project",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": " ",
          },
          Object {
            "type": 1,
            "value": "text",
          },
          Object {
            "type": 0,
            "value": " projects",
          },
        ],
      },
    },
    "pluralType": "cardinal",
    "type": 6,
    "value": "c",
  },
]
`;

exports[`parse() can parse '{count, time}' 1`] = `
Array [
  Object {
    "style": null,
    "type": 4,
    "value": "count",
  },
]
`;

exports[`parse() can parse '{floor, selectordinal, =0{ground} one{#st} two{#nd} few{#rd} other{#th}} floor' 1`] = `
Array [
  Object {
    "offset": 0,
    "options": Object {
      "=0": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "ground",
          },
        ],
      },
      "few": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "floor",
          },
          Object {
            "type": 0,
            "value": "rd",
          },
        ],
      },
      "one": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "floor",
          },
          Object {
            "type": 0,
            "value": "st",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "floor",
          },
          Object {
            "type": 0,
            "value": "th",
          },
        ],
      },
      "two": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "floor",
          },
          Object {
            "type": 0,
            "value": "nd",
          },
        ],
      },
    },
    "pluralType": "ordinal",
    "type": 6,
    "value": "floor",
  },
  Object {
    "type": 0,
    "value": " floor",
  },
]
`;

exports[`parse() can parse '{gender, select, female {woman} male {man} other {person}}' 1`] = `
Array [
  Object {
    "options": Object {
      "female": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "woman",
          },
        ],
      },
      "male": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "man",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "person",
          },
        ],
      },
    },
    "type": 5,
    "value": "gender",
  },
]
`;

exports[`parse() can parse '{num, number, percent}' 1`] = `
Array [
  Object {
    "style": "percent",
    "type": 2,
    "value": "num",
  },
]
`;

exports[`parse() can parse '{numPhotos, plural, =0{no photos} =1{one photo} other{# photos}}' 1`] = `
Array [
  Object {
    "offset": 0,
    "options": Object {
      "=0": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "no photos",
          },
        ],
      },
      "=1": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "one photo",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "numPhotos",
          },
          Object {
            "type": 0,
            "value": " photos",
          },
        ],
      },
    },
    "pluralType": "cardinal",
    "type": 6,
    "value": "numPhotos",
  },
]
`;

exports[`parse() can parse 'Foo {var1, plural, =0{# var1} other{{var2, plural, =0{# var2} other{# var2-other}} # other}}' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "Foo ",
  },
  Object {
    "offset": 0,
    "options": Object {
      "=0": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "var1",
          },
          Object {
            "type": 0,
            "value": " var1",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "offset": 0,
            "options": Object {
              "=0": Object {
                "location": undefined,
                "value": Array [
                  Object {
                    "style": null,
                    "type": 2,
                    "value": "var2",
                  },
                  Object {
                    "type": 0,
                    "value": " var2",
                  },
                ],
              },
              "other": Object {
                "location": undefined,
                "value": Array [
                  Object {
                    "style": null,
                    "type": 2,
                    "value": "var2",
                  },
                  Object {
                    "type": 0,
                    "value": " var2-other",
                  },
                ],
              },
            },
            "pluralType": "cardinal",
            "type": 6,
            "value": "var2",
          },
          Object {
            "type": 0,
            "value": " ",
          },
          Object {
            "style": null,
            "type": 2,
            "value": "var1",
          },
          Object {
            "type": 0,
            "value": " other",
          },
        ],
      },
    },
    "pluralType": "cardinal",
    "type": 6,
    "value": "var1",
  },
]
`;

exports[`parse() can parse 'Hello, {name}!' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "Hello, ",
  },
  Object {
    "type": 1,
    "value": "name",
  },
  Object {
    "type": 0,
    "value": "!",
  },
]
`;

exports[`parse() can parse 'Hello, World!' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "Hello, World!",
  },
]
`;

exports[`parse() can parse 'My name is {FIRST} {LAST}, age {age, number}, time {time, time}, date {date, date}.' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "My name is ",
  },
  Object {
    "type": 1,
    "value": "FIRST",
  },
  Object {
    "type": 0,
    "value": " ",
  },
  Object {
    "type": 1,
    "value": "LAST",
  },
  Object {
    "type": 0,
    "value": ", age ",
  },
  Object {
    "style": null,
    "type": 2,
    "value": "age",
  },
  Object {
    "type": 0,
    "value": ", time ",
  },
  Object {
    "style": null,
    "type": 4,
    "value": "time",
  },
  Object {
    "type": 0,
    "value": ", date ",
  },
  Object {
    "style": null,
    "type": 3,
    "value": "date",
  },
  Object {
    "type": 0,
    "value": ".",
  },
]
`;

exports[`parse() can parse 'This '{isn''t}' obvious' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "This {isn't} obvious",
  },
]
`;

exports[`parse() can parse 'this is {count,plural,offset:1 one{{count, number} dog} other{{count, number} dogs}}' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "this is ",
  },
  Object {
    "offset": 1,
    "options": Object {
      "one": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "count",
          },
          Object {
            "type": 0,
            "value": " dog",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "count",
          },
          Object {
            "type": 0,
            "value": " dogs",
          },
        ],
      },
    },
    "pluralType": "cardinal",
    "type": 6,
    "value": "count",
  },
]
`;

exports[`parse() can print AST from '   some random test   ' 1`] = `"   some random test   "`;

exports[`parse() can print AST from ''#'' 1`] = `"'#'"`;

exports[`parse() can print AST from ''{'' 1`] = `"'{'"`;

exports[`parse() can print AST from ''{name}'' 1`] = `"'{name}'"`;

exports[`parse() can print AST from ''}'' 1`] = `"'}'"`;

exports[`parse() can print AST from '{  num , number,percent  }' 1`] = `"{num, number, percent}"`;

exports[`parse() can print AST from '{c, plural, =1 { {text} project} other { {text} projects}}' 1`] = `"{c,plural,=1{ {text} project} other{ {text} projects}}"`;

exports[`parse() can print AST from '{c, plural, =99 { {text} project} other { {text} projects}}' 1`] = `"{c,plural,=99{ {text} project} other{ {text} projects}}"`;

exports[`parse() can print AST from '{count, time}' 1`] = `"{count, time}"`;

exports[`parse() can print AST from '{floor, selectordinal, =0{ground} one{#st} two{#nd} few{#rd} other{#th}} floor' 1`] = `"{floor,selectordinal,=0{ground} one{{floor, number}st} two{{floor, number}nd} few{{floor, number}rd} other{{floor, number}th}} floor"`;

exports[`parse() can print AST from '{gender, select, female {woman} male {man} other {person}}' 1`] = `"{gender,select,female{woman} male{man} other{person}}"`;

exports[`parse() can print AST from '{num, number, percent}' 1`] = `"{num, number, percent}"`;

exports[`parse() can print AST from '{numPhotos, plural, =0{no photos} =1{one photo} other{# photos}}' 1`] = `"{numPhotos,plural,=0{no photos} =1{one photo} other{{numPhotos, number} photos}}"`;

exports[`parse() can print AST from 'Foo {var1, plural, =0{# var1} other{{var2, plural, =0{# var2} other{# var2-other}} # other}}' 1`] = `"Foo {var1,plural,=0{{var1, number} var1} other{{var2,plural,=0{{var2, number} var2} other{{var2, number} var2-other}} {var1, number} other}}"`;

exports[`parse() can print AST from 'Hello, {name}!' 1`] = `"Hello, {name}!"`;

exports[`parse() can print AST from 'Hello, World!' 1`] = `"Hello, World!"`;

exports[`parse() can print AST from 'My name is {FIRST} {LAST}, age {age, number}, time {time, time}, date {date, date}.' 1`] = `"My name is {FIRST} {LAST}, age {age, number}, time {time, time}, date {date, date}."`;

exports[`parse() can print AST from 'This '{isn''t}' obvious' 1`] = `"This '{isn't}' obvious"`;

exports[`parse() can print AST from 'this is {count,plural,offset:1 one{{count, number} dog} other{{count, number} dogs}}' 1`] = `"this is {count,plural,offset:1 one{{count, number} dog} other{{count, number} dogs}}"`;

exports[`parse({ captureLocation: true }) can parse '   some random test   ' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "   some random test   ",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse ''#'' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "'#'",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse ''{'' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "{",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse ''{name}'' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "{name}",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse ''}'' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "}",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse '{  num , number,percent  }' 1`] = `
Array [
  Object {
    "style": "percent",
    "type": 2,
    "value": "num",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse '{c, plural, =1 { {text} project} other { {text} projects}}' 1`] = `
Array [
  Object {
    "offset": 0,
    "options": Object {
      "=1": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": " ",
          },
          Object {
            "type": 1,
            "value": "text",
          },
          Object {
            "type": 0,
            "value": " project",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": " ",
          },
          Object {
            "type": 1,
            "value": "text",
          },
          Object {
            "type": 0,
            "value": " projects",
          },
        ],
      },
    },
    "pluralType": "cardinal",
    "type": 6,
    "value": "c",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse '{c, plural, =99 { {text} project} other { {text} projects}}' 1`] = `
Array [
  Object {
    "offset": 0,
    "options": Object {
      "=99": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": " ",
          },
          Object {
            "type": 1,
            "value": "text",
          },
          Object {
            "type": 0,
            "value": " project",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": " ",
          },
          Object {
            "type": 1,
            "value": "text",
          },
          Object {
            "type": 0,
            "value": " projects",
          },
        ],
      },
    },
    "pluralType": "cardinal",
    "type": 6,
    "value": "c",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse '{count, time}' 1`] = `
Array [
  Object {
    "style": null,
    "type": 4,
    "value": "count",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse '{floor, selectordinal, =0{ground} one{#st} two{#nd} few{#rd} other{#th}} floor' 1`] = `
Array [
  Object {
    "offset": 0,
    "options": Object {
      "=0": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "ground",
          },
        ],
      },
      "few": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "floor",
          },
          Object {
            "type": 0,
            "value": "rd",
          },
        ],
      },
      "one": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "floor",
          },
          Object {
            "type": 0,
            "value": "st",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "floor",
          },
          Object {
            "type": 0,
            "value": "th",
          },
        ],
      },
      "two": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "floor",
          },
          Object {
            "type": 0,
            "value": "nd",
          },
        ],
      },
    },
    "pluralType": "ordinal",
    "type": 6,
    "value": "floor",
  },
  Object {
    "type": 0,
    "value": " floor",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse '{gender, select, female {woman} male {man} other {person}}' 1`] = `
Array [
  Object {
    "options": Object {
      "female": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "woman",
          },
        ],
      },
      "male": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "man",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "person",
          },
        ],
      },
    },
    "type": 5,
    "value": "gender",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse '{num, number, percent}' 1`] = `
Array [
  Object {
    "style": "percent",
    "type": 2,
    "value": "num",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse '{numPhotos, plural, =0{no photos} =1{one photo} other{# photos}}' 1`] = `
Array [
  Object {
    "offset": 0,
    "options": Object {
      "=0": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "no photos",
          },
        ],
      },
      "=1": Object {
        "location": undefined,
        "value": Array [
          Object {
            "type": 0,
            "value": "one photo",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "numPhotos",
          },
          Object {
            "type": 0,
            "value": " photos",
          },
        ],
      },
    },
    "pluralType": "cardinal",
    "type": 6,
    "value": "numPhotos",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse 'Foo {var1, plural, =0{# var1} other{{var2, plural, =0{# var2} other{# var2-other}} # other}}' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "Foo ",
  },
  Object {
    "offset": 0,
    "options": Object {
      "=0": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "var1",
          },
          Object {
            "type": 0,
            "value": " var1",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "offset": 0,
            "options": Object {
              "=0": Object {
                "location": undefined,
                "value": Array [
                  Object {
                    "style": null,
                    "type": 2,
                    "value": "var2",
                  },
                  Object {
                    "type": 0,
                    "value": " var2",
                  },
                ],
              },
              "other": Object {
                "location": undefined,
                "value": Array [
                  Object {
                    "style": null,
                    "type": 2,
                    "value": "var2",
                  },
                  Object {
                    "type": 0,
                    "value": " var2-other",
                  },
                ],
              },
            },
            "pluralType": "cardinal",
            "type": 6,
            "value": "var2",
          },
          Object {
            "type": 0,
            "value": " ",
          },
          Object {
            "style": null,
            "type": 2,
            "value": "var1",
          },
          Object {
            "type": 0,
            "value": " other",
          },
        ],
      },
    },
    "pluralType": "cardinal",
    "type": 6,
    "value": "var1",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse 'Hello, {name}!' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "Hello, ",
  },
  Object {
    "type": 1,
    "value": "name",
  },
  Object {
    "type": 0,
    "value": "!",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse 'Hello, World!' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "Hello, World!",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse 'My name is {FIRST} {LAST}, age {age, number}, time {time, time}, date {date, date}.' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "My name is ",
  },
  Object {
    "type": 1,
    "value": "FIRST",
  },
  Object {
    "type": 0,
    "value": " ",
  },
  Object {
    "type": 1,
    "value": "LAST",
  },
  Object {
    "type": 0,
    "value": ", age ",
  },
  Object {
    "style": null,
    "type": 2,
    "value": "age",
  },
  Object {
    "type": 0,
    "value": ", time ",
  },
  Object {
    "style": null,
    "type": 4,
    "value": "time",
  },
  Object {
    "type": 0,
    "value": ", date ",
  },
  Object {
    "style": null,
    "type": 3,
    "value": "date",
  },
  Object {
    "type": 0,
    "value": ".",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse 'This '{isn''t}' obvious' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "This {isn't} obvious",
  },
]
`;

exports[`parse({ captureLocation: true }) can parse 'this is {count,plural,offset:1 one{{count, number} dog} other{{count, number} dogs}}' 1`] = `
Array [
  Object {
    "type": 0,
    "value": "this is ",
  },
  Object {
    "offset": 1,
    "options": Object {
      "one": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "count",
          },
          Object {
            "type": 0,
            "value": " dog",
          },
        ],
      },
      "other": Object {
        "location": undefined,
        "value": Array [
          Object {
            "style": null,
            "type": 2,
            "value": "count",
          },
          Object {
            "type": 0,
            "value": " dogs",
          },
        ],
      },
    },
    "pluralType": "cardinal",
    "type": 6,
    "value": "count",
  },
]
`;

exports[`parse({ captureLocation: true }) can print AST from '   some random test   ' 1`] = `"   some random test   "`;

exports[`parse({ captureLocation: true }) can print AST from ''#'' 1`] = `"'#'"`;

exports[`parse({ captureLocation: true }) can print AST from ''{'' 1`] = `"'{'"`;

exports[`parse({ captureLocation: true }) can print AST from ''{name}'' 1`] = `"'{name}'"`;

exports[`parse({ captureLocation: true }) can print AST from ''}'' 1`] = `"'}'"`;

exports[`parse({ captureLocation: true }) can print AST from '{  num , number,percent  }' 1`] = `"{num, number, percent}"`;

exports[`parse({ captureLocation: true }) can print AST from '{c, plural, =1 { {text} project} other { {text} projects}}' 1`] = `"{c,plural,=1{ {text} project} other{ {text} projects}}"`;

exports[`parse({ captureLocation: true }) can print AST from '{c, plural, =99 { {text} project} other { {text} projects}}' 1`] = `"{c,plural,=99{ {text} project} other{ {text} projects}}"`;

exports[`parse({ captureLocation: true }) can print AST from '{count, time}' 1`] = `"{count, time}"`;

exports[`parse({ captureLocation: true }) can print AST from '{floor, selectordinal, =0{ground} one{#st} two{#nd} few{#rd} other{#th}} floor' 1`] = `"{floor,selectordinal,=0{ground} one{{floor, number}st} two{{floor, number}nd} few{{floor, number}rd} other{{floor, number}th}} floor"`;

exports[`parse({ captureLocation: true }) can print AST from '{gender, select, female {woman} male {man} other {person}}' 1`] = `"{gender,select,female{woman} male{man} other{person}}"`;

exports[`parse({ captureLocation: true }) can print AST from '{num, number, percent}' 1`] = `"{num, number, percent}"`;

exports[`parse({ captureLocation: true }) can print AST from '{numPhotos, plural, =0{no photos} =1{one photo} other{# photos}}' 1`] = `"{numPhotos,plural,=0{no photos} =1{one photo} other{{numPhotos, number} photos}}"`;

exports[`parse({ captureLocation: true }) can print AST from 'Foo {var1, plural, =0{# var1} other{{var2, plural, =0{# var2} other{# var2-other}} # other}}' 1`] = `"Foo {var1,plural,=0{{var1, number} var1} other{{var2,plural,=0{{var2, number} var2} other{{var2, number} var2-other}} {var1, number} other}}"`;

exports[`parse({ captureLocation: true }) can print AST from 'Hello, {name}!' 1`] = `"Hello, {name}!"`;

exports[`parse({ captureLocation: true }) can print AST from 'Hello, World!' 1`] = `"Hello, World!"`;

exports[`parse({ captureLocation: true }) can print AST from 'My name is {FIRST} {LAST}, age {age, number}, time {time, time}, date {date, date}.' 1`] = `"My name is {FIRST} {LAST}, age {age, number}, time {time, time}, date {date, date}."`;

exports[`parse({ captureLocation: true }) can print AST from 'This '{isn''t}' obvious' 1`] = `"This '{isn't}' obvious"`;

exports[`parse({ captureLocation: true }) can print AST from 'this is {count,plural,offset:1 one{{count, number} dog} other{{count, number} dogs}}' 1`] = `"this is {count,plural,offset:1 one{{count, number} dog} other{{count, number} dogs}}"`;
